File: Regression/Polynomial Linear Regression/polynomial_regression.r
#Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

#Fitting Linear Regression to dataset
lin_reg = lm(formula =Salary ~ . ,data = dataset )

#Fitting Polynomial Regression to dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ . , data = dataset)

#Visualizing Linear Regression Results
#install.packages(ggplot2)
library(ggplot2)
ggplot()+
  geom_point(aes(x =dataset$Level , y = dataset$Salary),
             colour = 'red')+
  geom_line(aes(x =dataset$Level , y = predict(lin_reg,newdata = dataset)),
            colour = 'blue')+
  ggtitle('Truth or Bluff(Linear Regression)')+
  xlab('Level ')+
  ylab('Salary ')

#Visualizing Polynomial Regression Results
library(ggplot2)
ggplot()+
  geom_point(aes(x =dataset$Level , y = dataset$Salary),
             colour = 'red')+
  geom_line(aes(x =dataset$Level , y = predict(poly_reg,newdata = dataset)),
            colour = 'blue')+
  ggtitle('Truth or Bluff(Polynomial Regression)')+
  xlab('Level ')+
  ylab('Salary ')

#Predicting new result with Linear Regression 
y_pred = predict(lin_reg,newdata = data.frame(Level = 6.5)) #in R you need data to be present in dataset to predict hence we add data by dataframe
#By adding 6.5 to Level column

#Predicting new result with Polynomial Regression
y_pred1 = predict(poly_reg,newdata = data.frame(Level = 6.5,
                                               Level2 = 6.5^2,
                                               Level3 = 6.5^3,
                                               Level4 = 6.5^4))

File: Regression/Polynomial Linear Regression/polynomial_regression.py
#Polynomial Regression 
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values #no need to include name of levels #1:2 instead of 1 to make sure we get matrix(10,1) and not array (10,)
y = dataset.iloc[:, 2].values

#Here we need all the levels in training set hence we wont divide it into training and test set
# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

#Fitting Linear Regression to dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

#Fitting Polynomial Regression to dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
poly_reg.fit(X_poly,y)
#now we have to fit polynomial features in multiple linear model
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly,y)

#Visualizing the Linear Regression Results
plt.scatter(X, y, color = 'red')
plt.plot(X , lin_reg.predict(X),color = 'blue')
plt.title('Truth or Bluff(Linear Regression Model)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

#Visualizing the Polynomial Regression Results
#X_grid = np.arange(min(X),max(X),0.1) #divides dataset 1-10 to 1,1.1,1.2,.......,9.9,10.0 for more accurate plot
#X_grid = X_grid.reshape(len(X_grid),1) #after this just replace X by X_grid in plot method
plt.scatter(X, y, color = 'red')
plt.plot(X , lin_reg2.predict(poly_reg.fit_transform(X)),color = 'blue')
plt.title('Truth or Bluff(Polynomial Regression Model)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

#Predicting new result with Linear Regression
lin_reg.predict(6.5)

#Predicting new result with Polynomial Regression
lin_reg2.predict(poly_reg.fit_transform(6.5))